### PR TITLE
Addon-Docs: Change URL hash when TOC item is clicked and when scrolling

### PR DESCRIPTION
**Pull request by sookmax**
## What I did

Hello, I've noticed nothing actually happens to the url when the item of the table of contents for a docs page (if enabled) is clicked. 

I thought it would be better for user experience if the current url actually changes to include the corresponding hash when a toc item is clicked. This PR includes that behavior.

While I was on it, I also added a behavior where the url hash changes reactively to the current scroll position and the current active link.

I've recorded short videos demonstrating the behaviors described above.

### Url hash change on click

https://github.com/storybookjs/storybook/assets/71210554/a5c10ddc-6a95-4350-ae08-ac2e95e08743

### Url hash change on scroll

https://github.com/storybookjs/storybook/assets/71210554/62a5dfe5-0a23-4671-b8dd-30731dba5432

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "build", "documentation", "maintenance", "dependencies", "other"]`

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: code/lib/router/src/router.tsx [hunk 1]
```diff
@@ -48,7 +48,12 @@ export const useNavigate = () => {
 
   return useCallback((to: R.To | number, { plain, ...options } = {} as NavigateOptions) => {
     if (typeof to === 'string' && to.startsWith('#')) {
-      document.location.hash = to;
+      if (!to.substring(1)) {
+        // if only '#' symbol is found, remove the hash entirely.
```
Comment (ndelangen) [inline code/lib/router/src/router.tsx:52]:
```suggestion
      if (to === '#') {
```

Would this be doing the same? It seems easier to understand?

**Comment by ndelangen** [comment]
This is great @sookmax ! I've noticed some really amazing PRs from you ❤️

File: code/lib/router/src/router.tsx (hunk 1 above)
Comment (sookmax) [reply code/lib/router/src/router.tsx:52]: Yes indeed! That's much better.

File: code/ui/blocks/src/components/TableOfContents.tsx [hunk 2]
```diff
@@ -152,7 +156,31 @@ export const TableOfContents = ({
...
        */
-      onClick: () => false,
+      onClick: (e: MouseEvent) => {
+        e.preventDefault();
+        if (e.currentTarget instanceof HTMLAnchorElement) {
+          const [, headerId] = e.currentTarget.href.split('#');
+          channel.emit(NAVIGATE_URL, `#${headerId}`);
+        }
+      },
+      scrollEndCallback: () => {
```
[the remark this answers is not in the document]
Comment (sookmax) [reply code/ui/blocks/src/components/TableOfContents.tsx:166]: Ah, that's what I hoped as well. I didn't want to use `querySelector` myself, but well, the sole argument that's passed to `scrollEndCallback` is an event object from `scroll` and `resize` events 😢.

https://github.com/tscanlin/tocbot/blob/9ab298453ceebd2bf47968a73723c0539ad42c89/src/js/index.js#L196-L214

[the remark this answers is not in the document]
Comment (ndelangen) [reply code/ui/blocks/src/components/TableOfContents.tsx:166]: Ok, thanks for checking!

**Review by JReinhold — CHANGES_REQUESTED** [review]
I like the behavior of changing the hash in the URL when a header-link is clicked, this is a pattern we see a lot of places.

I'm not a big fan of changing the hash automatically on scroll. I've never seen this UX pattern before and as a user I'd be surprised if this happened.

Thoughts @cdedreuille ?

**Comment by Sidnioulz** [comment]
> I like the behavior of changing the hash in the URL when a header-link is clicked, this is a pattern we see a lot of places.
> 
> I'm not a big fan of changing the hash automatically on scroll. I've never seen this UX pattern before and as a user I'd be surprised if this happened.
> 
> Thoughts @cdedreuille ?

@JReinhold That's something I see happen every now and then. I personally like it: the moment where I share a doc page to a third-party is likely the moment where I've encountered the piece of info I've been looking for. If the URL already links to the relevant section, it's likely faster for my third-party to locate the information in the document.

There's not much relevant contnent on the UX Stack exchange: https://ux.stackexchange.com/questions/30361/how-to-handle-history-events-from-scolling-and-other-in-page-events/30663#30663. And not all the answers on it are on-topic, sadly.

A quick look at major documentation tools shows they do not auto-inject the hash on scroll, so adopting the majority behaviour in the absence of hard UR data would mean not changing the hash.

@sookmax I'm happy to finalise the request changes for you if you no longer have time for it.

**Comment by sookmax** [comment]
> @sookmax I'm happy to finalise the request changes for you if you no longer have time for it.

Thanks @Sidnioulz. Feel free to pick up where I left off.

On automatic hash change on scroll: I'm pretty sure [Vitepress](https://vitepress.dev/) had had this behavior when I created this doc: [example](https://tour3d-v1.vercel.app/guide/getting-started.html). But I don't think that's the default case anymore, judging from their official docs' scrolling behavior.

File: code/ui/blocks/src/components/TableOfContents.tsx
Comment (Sidnioulz) [inline code/ui/blocks/src/components/TableOfContents.tsx:1]: There's something fishy about the useEffect dependencies both in the original and proposed code.

In the original code, we initialise tocbot even when the ToC is disabled, which is wasteful. @JReinhold should I use this opportunity to ensure the hook has no effects when the ToC is hidden?

In the newly proposed code, the SB channel and several new props are consumed by the event handlers being installed in the tocbot. Therefore, they must be included in the dependencies for tocbot to work as intended.

**Comment by Sidnioulz** [comment]
> On automatic hash change on scroll: I'm pretty sure [Vitepress](https://vitepress.dev/) had had this behavior when I created this doc: [example](https://tour3d-v1.vercel.app/guide/getting-started.html). But I don't think that's the default case anymore, judging from their official docs' scrolling behavior.

Yeah I remember it that way too. It surprised me to see it had changed.

As I can't push directly to your branch, I've forked it and pushed it to https://github.com/storybookjs/storybook/pull/30130.

**Comment by Sidnioulz** [comment]
@JReinhold @sookmax I'm taking the liberty to close this one now that the other one, rebased and with the requested changes and extra bug fixes, is open: https://github.com/storybookjs/storybook/pull/30130

**Comment by sookmax** [comment]
Thanks for picking this up and finishing it @Sidnioulz, it is very much appreciated.

**Pull request closed without merging**